### PR TITLE
Split bundle_skel.sh into bundle_skel_<PLATFORM>.sh

### PR DESCRIPTION
**Pull request by kapilash**
bundle_skel.sh is split into
* installer/bundle/bundle_skel_AIX.sh
* installer/bundle/bundle_skel_HPUX.sh
* installer/bundle/bundle_skel_Linux.sh
* installer/bundle/bundle_skel_SunOS.sh

Change Linux bundle to install/upgrade all packages as a set

### CODE REVIEW

File: installer/bundle/bundle_skel_AIX.sh [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    *)
+        PWD="`pwd`"
+        SCRIPT_INDIRECT="`dirname $PWD/$0`"
+        ;;
+esac
+
+SCRIPT_DIR="`(cd \"$SCRIPT_INDIRECT\"; pwd -P)`"
+SCRIPT="$SCRIPT_DIR/`basename $0`"
+EXTRACT_DIR="`pwd -P`/scxbundle.$$"
+DPKG_CONF_QUALS="--force-confold --force-confdef"
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Not needed on AIX

Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: DPKG qualifiers won't be needed on AIX

File: installer/bundle/bundle_skel_AIX.sh [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+# These symbols will get replaced during the bundle creation process.
+#
+# The OM_PKG symbol should contain something like:
+#       scx-1.5.1-115.rhel.6.x64 (script adds .rpm or .deb, as appropriate)
+# Note that for non-Linux platforms, this symbol should contain full filename.
+#
+# PROVIDER_ONLY is normally set to '0'. Set to non-zero if you wish to build a
+# version of SCX that is only the provider (no OMI, no bundled packages). This
+# essentially provides a "scx-cimprov" type package if just the provider alone
+# must be included as part of some other package.
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: First, PROVIDER_ONLY was only done for Linux. So you can remove this from all UNIX scripts (and there are lots of 'if' tests related to this).

Second, it turns out that PROVIDER_ONLY isn't really used. This was added for OMS, when we were thinking that SCX would just be bundled as a PROVIDER_ONLY shell bundle. But it turns out that OMS just builds SCX into it's own bundle (building it). Thus, it turns out that PROVIDER_ONLY isn't even needed on Linux.

I suggest you remove this, as it removes considerable cruft (certainly from UNIX bundles, but really from all bundles, Linux too).

Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: First, PROVIDER_ONLY was only done for Linux. So you can remove this from all UNIX scripts (and there are lots of 'if' tests related to this).

Second, it turns out that PROVIDER_ONLY isn't really used. This was added for OMS, when we were thinking that SCX would just be bundled as a PROVIDER_ONLY shell bundle. But it turns out that OMS just builds SCX into it's own bundle (building it). Thus, it turns out that PROVIDER_ONLY isn't even needed on Linux.

I suggest you remove this, as it removes considerable cruft (certainly from UNIX bundles, but really from all bundles, Linux too).

File: installer/bundle/bundle_skel_AIX.sh [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+
+TAR_FILE=<TAR_FILE>
+OM_PKG=<OM_PKG>
+OMI_PKG=<OMI_PKG>
+PROVIDER_ONLY=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: See above. This can be removed. I imagine this is in all of the bundle files; I won't mention this again for the other bundle files, but please do remove this (and associated 'if' tests everywhere).

File: installer/bundle/bundle_skel_AIX.sh [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+}
+
+# $1 - The filename of the package to be installed
+# $2 - The package name of the package to be installed
+pkg_add() {
+    pkg_filename=$1
+    pkg_name=$2
+
+    echo "----- Installing package: $pkg_name ($pkg_filename) -----"
+    /usr/sbin/installp -a -X -d $pkg_filename ${pkg_name}.rte
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Wow, that's a whole lot simpler when it's not testing for 'n' different platforms!

File: installer/bundle/bundle_skel_AIX.sh [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+    pkg_allowed=$3
+
+    echo "----- Updating package: $pkg_name ($pkg_filename) -----"
+
+    if [ -z "${forceFlag}" -a -n "$pkg_allowed" ]; then
+        if [ $pkg_allowed -ne 0 ]; then
+            echo "Skipping package since existing version >= version available"
+            return 0
+        fi
+    fi
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: The notion of "skipping packages" is all based around version detection for Azure, which isn't appropriate for non-Linux platforms. Please verify how this is called and, if so, remove this 'if' condition. Makes pkg_upd about as simple as pkg_add.

File: installer/bundle/bundle_skel_AIX.sh [hunk 6]
```diff
@@ -0,0 +1,432 @@
...
+case "$installMode" in
+    E)
+        # Files are extracted, so just exit
+        cleanup_and_exit 0 "SAVE"
+        ;;
+
+    I)
+        echo "Installing cross-platform agent ..."
+
+        if [ $PROVIDER_ONLY -eq 0 ]; then
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: This is one of the conditions that should be able to go away. When you get rid of this, be sure to fix indentation of the code below this (within the 'if' block). If you use emacs, it's wicked easy to do this ('indent-region').

File: installer/bundle/bundle_skel_AIX.sh [hunk 7]
```diff
@@ -0,0 +1,432 @@
...
+                [ ! -f $i ] && break
+                ./$i
+                if [ $? -eq 0 ]; then
+                    OSS_BUNDLE=`basename $i -oss-test.sh`
+                    ./${OSS_BUNDLE}-cimprov-*.sh --install $FORCE $restartDependencies
+                    TEMP_STATUS=$?
+                    [ $TEMP_STATUS -ne 0 ] && BUNDLE_EXIT_STATUS="$TEMP_STATUS"
+                fi
+            done
+        fi
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: The notion of "mega-bundles" (bundles that install bundles) is a linux-only thing. You should be able to remove 366-380 here, and similar code for all other UNIX-only installers.

I'm being pedantic because now's our chance to get these shell bundles MUCH cleaner than they've ever been before, and thus much easier to maintain.

Make similar changes for HP and Solaris.

File: installer/bundle/bundle_skel_AIX.sh [hunk 8]
```diff
@@ -0,0 +1,432 @@
...
+        ;;
+
+    U)
+        echo "Updating cross-platform agent ..."
+        if [ $PROVIDER_ONLY -eq 0 ]; then
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Search for all occurrences of $PROVIDER_ONLY and remove in all UNIX scripts (but not Linux).

File: installer/bundle/bundle_skel_HPUX.sh [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+    *)
+        PWD="`pwd`"
+        SCRIPT_INDIRECT="`dirname $PWD/$0`"
+        ;;
+esac
+
+SCRIPT_DIR="`(cd \"$SCRIPT_INDIRECT\"; pwd -P)`"
+SCRIPT="$SCRIPT_DIR/`basename $0`"
+EXTRACT_DIR="`pwd -P`/scxbundle.$$"
+DPKG_CONF_QUALS="--force-confold --force-confdef"
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_HPUX.sh]: Ditto for AIX script, won't bring this up again for further scripts.

File: installer/bundle/bundle_skel_AIX.sh [hunk 10]
```diff
@@ -0,0 +1,432 @@
...
+        /opt/microsoft/scx/bin/uninstall $installMode
+    else
+        # This is an old kit.  Let's remove each separate provider package
+        for i in /opt/microsoft/*-cimprov; do
+            PKG_NAME=`basename $i`
+            if [ "$PKG_NAME" != "*-cimprov" ]; then
+                echo "Removing ${PKG_NAME} ..."
+                pkg_rm ${PKG_NAME}
+            fi
+        done
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Lines 280 - 286 is to remove bundled packages. Since UNIX doesn't have that, you can remove this in all UNIX-based installation scripts. For UNIX, just removing scx and OMI (done at 289-290) should be fine.

File: installer/bundle/bundle_skel_AIX.sh [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+        # Now just simply pkg_rm scx (and omi if it has no dependencies)
+        pkg_rm scx
+        pkg_rm omi
+    fi
+
+    if [ "$installMode" = "P" ]
+    then
+        echo "Purging all files in cross-platform agent ..."
+        rm -rf /etc/opt/microsoft/*-cimprov /etc/opt/microsoft/scx /opt/microsoft/*-cimprov /opt/microsoft/scx /var/opt/microsoft/*-cimprov /var/opt/microsoft/scx
+        rmdir /etc/opt/microsoft /opt/microsoft /var/opt/microsoft 1>/dev/null 2>/dev/null
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: You can get rid of all *-cimprov constructs since that can never exist on UNIX.

File: installer/bundle/bundle_skel_AIX.sh [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+                [ ! -f $i ] && break
+                ./$i
+                if [ $? -eq 0 ]; then
+                    OSS_BUNDLE=`basename $i -oss-test.sh`
+                    ./${OSS_BUNDLE}-cimprov-*.sh --upgrade $FORCE $restartDependencies
+                    TEMP_STATUS=$?
+                    [ $TEMP_STATUS -ne 0 ] && BUNDLE_EXIT_STATUS="$TEMP_STATUS"
+                fi
+            done
+        fi
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: No mega-bundles for UNIX, lines 400-416 can go away here (and similar code in all other UNIX-based files).

File: installer/bundle/bundle_skel_AIX.sh [hunk 13]
```diff
@@ -0,0 +1,432 @@
...
+fi
+
+#
+# Do stuff after extracting the binary here, such as actually installing the package.
+#
+
+EXIT_STATUS=0
+SCX_EXIT_STATUS=0
+OMI_EXIT_STATUS=0
+BUNDLE_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: I imagine, since bundles don't exist on UNIX, that BUNDLE_EXIT_STATUS won't exist either when you clean up the code below.

File: installer/bundle/bundle_skel_AIX.sh [hunk 14]
```diff
@@ -0,0 +1,432 @@
...
+        ;;
+
+    *)
+        echo "$0: Invalid setting of variable \$installMode, exiting" >&2
+        cleanup_and_exit 2
+esac
+
+# Remove temporary files (now part of cleanup_and_exit) and exit
+
+if [ "$SCX_EXIT_STATUS" -ne 0 -o "$OMI_EXIT_STATUS" -ne 0 -o "$BUNDLE_EXIT_STATUS" -ne 0 ]; then
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Should be able to remove BUNDLE_EXIT_STATUS from this list of tests for all UNIX-based files.

File: installer/bundle/bundle_skel_AIX.sh [hunk 15]
```diff
@@ -0,0 +1,432 @@
...
+                # It is acceptable that this fails due to the new omi being 
+                # the same version (or less) than the one currently installed.
+                OMI_EXIT_STATUS=0
+            else
+                pkg_add $OMI_PKG omi
+                OMI_EXIT_STATUS=$?  
+            fi
+        fi
+
+        pkg_upd $OM_PKG scx $?
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: What's $? here? I think you got rid of code above (probably calling version checking code), definitely get rid of it here, could cause odd failures.

File: installer/bundle/bundle_skel_AIX.sh [hunk 16]
```diff
@@ -0,0 +1,432 @@
...
+            done
+        fi
+        ;;
+
+    U)
+        echo "Updating cross-platform agent ..."
+        if [ $PROVIDER_ONLY -eq 0 ]; then
+            check_if_pkg_is_installed omi
+            if [ $? -eq 0 ]; then
+                pkg_upd $OMI_PKG omi $?
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: What's $? here? I think you got rid of code above (probably calling version checking code), definitely get rid of it here, could cause odd failures.

File: installer/bundle/bundle_skel_AIX.sh [hunk 17]
```diff
@@ -0,0 +1,432 @@
...
+}
+
+
+# $1 - The filename of the package to be installed
+# $2 - The package name of the package to be installed
+# $3 - Okay to upgrade the package? (Optional)
+pkg_upd() {
+    pkg_filename=$1
+    pkg_name=$2
+    pkg_allowed=$3
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: It looks like, if you get rid of code just below, you no longer need pkg_allowed at all. When you fix up calling code (get rid of stray `$?` parameters), you'll never call pkg_upd with three parameters.

File: installer/bundle/bundle_skel_AIX.sh [hunk 18]
```diff
@@ -0,0 +1,432 @@
...
+            ;;
+
+        --upgrade)
+            verifyNoInstallationOption
+            installMode=U
+            shift 1
+            ;;
+
+        --version)
+            echo "Version: `getVersionNumber $OM_PKG scx-`"
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Hmm, this was intended as a test helper when I was doing version detection code for Linux. Does this even work properly for UNIX platforms? If it does, leave it. But if it doesn't, just remove it for all UNIX platforms along with the getVersionNumber function. Leave for Linux, of course.

If you prefer, you can fix it to work on all UNIX platforms. But if it doesn't already work, it likely isn't worth it.

File: installer/bundle/bundle_skel_Linux.sh [hunk 19]
```diff
@@ -702,29 +581,25 @@ case "$installMode" in
...
         fi
 
-        pkg_add $OM_PKG scx
-        SCX_EXIT_STATUS=$?
+        pkg_add_list $OM_PKG scx
+
+        pkg_upd ${UPD_PKG_QUEUE}
+        # It is acceptable that this fails due to the new omi being 
+        # the same version (or less) than the one currently installed.
+        OMI_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: This is kind of whacky. I know that this is how you got it - my bad.

The intention here, before my most recent changes, is that if OMI was already installed, don't error out for that. But now, with the queue, setting OMI_EXIT_STATUS to zero is just wrong (as it'll not catch SCX install issues).

I think, strictly, this will work (amazingly enough), given that if nothing gets added to the list at 588, then 596 will be ignored since the list is empty. But still, whacky.

My suggestion is just to test if it's installed (at 587) and, if it is, do the same version check that the regular upgrade path does. Then, right after 588, call pkg_upd to actually do the update (since all other code below uses pkg_add). But here, since you've done the pre-check, actually do signal an error in OMI_EXIT_STATUS (since an error might have occurred).

File: installer/bundle/bundle_skel_Linux.sh [hunk 20]
```diff
@@ -749,19 +624,20 @@ case "$installMode" in
...
             fi
         fi
 
         shouldInstall_scx
-        pkg_upd $OM_PKG scx $?
-        SCX_EXIT_STATUS=$?
+        pkg_upd_list $OM_PKG scx $?
+
+        pkg_add ${ADD_PKG_QUEUE}
+        OMI_EXIT_STATUS=$?
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: Same messyness as above. These two lines should probably go right after 629.

Frankly, I'm not entirely sure you should do a separate test here (installing initially via upgrade isn't an error for RPM). But if I were you, I'd just want to get this wrapped up, so don't bother removing the 'if' test at 625 if you don't want to. I'll leave that up to you.

**Comment by jeffaco** [comment]
Okay, there's lot of remaining cleanup here, particularly for UNIX platforms. I only reviewed the first UNIX file (AIX), assuming you'll make all related changes to the other UNIX files (HP, Solaris). And I want to use your time efficiently. So here's my proposal:

1. Clean up AIX and all other UNIX files,
2. Clean up Linux file,

Do a limited test pass. By that, I mean: Do a pbuild, and install on at least one flavor of each platform (AIX 6.1 OR AIX 7.1, don't bother doing both; one HP, one Solaris, and one Linux). Then send me an updated PR. I'll do one final pass and, assuming no remaining problem will sign off.

At that point, assuming no further changes, do a complete test pass (install on all platforms that we support) and commit. If any of the tests fail, then fix that platform and rerun all installation tests for that platform.

The idea here is that, for the final set of changes, you want to install and upgrade on every single platform at least once. But you don't need to do that if you know further changes are coming down the line. And I don't want to catch typos that at least a basic install / upgrade would have caught.

Seem reasonable? If so, do another cleaning pass and send me a new PR, thanks. Hopefully you can get to this quickly; I think we'd both like to wrap this up in the next day or so.

Nicely done, by the way. Having one combined file was a real mess after a while. I think these improvements will make a big difference in maintainability. Imagine just testing on Linux if that's all you changed! :smile:

File: installer/bundle/bundle_skel_AIX.sh (hunk 2 above)
Comment (kapilash) [inline installer/bundle/bundle_skel_AIX.sh]: > I suggest you remove this, as it removes considerable cruft (certainly from UNIX bundles, but really from all bundles, Linux too).

In [other comment](https://github.com/Microsoft/SCXcore/pull/33#discussion_r63546786), you said 
> Search for all occurrences of $PROVIDER_ONLY and remove in all UNIX scripts (but not Linux).

So for now I am keeping PROVIDER_ONLY for Linux. will remove it in the next iteration. (Testing will be limited to only Linux versions)

File: installer/bundle/bundle_skel_Linux.sh (hunk 19 above)
Comment (kapilash) [inline installer/bundle/bundle_skel_Linux.sh]: > setting OMI_EXIT_STATUS to zero is just wrong (as it'll not catch SCX install issues).

SCX install issues are not relevant here.. right?  the ```UPD_PKG_QUEUE``` would not contain scx at this point.  If the script is at this line, it means the user is running the script with ```--install``` option. So, UPD_PKG_QUEUE will either be empty or will only contain omi. 

> My suggestion is just to test if it's installed (at 587) and, if it is, do the same version check that the regular upgrade path does. Then, right after 588, call pkg_upd to actually do the update (since all other code below uses pkg_add). 

OK.

Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: Sorry for confusion. When I said "it'll not catch SCX install issues", I meant that more generically (since the shell bundle ultimately installs SCX with OMI).

Perhaps a more precise way to say what I was trying to say:

"Setting OMI_EXIT_STATUS to zero is just wrong (as it won't catch actual installation problems with the OMI component)."

But I think you figured out what I meant! :smile:

File: installer/bundle/bundle_skel_AIX.sh [hunk 21]
```diff
@@ -0,0 +1,373 @@
...
+
+    I)
+        echo "Installing cross-platform agent ..."
+
+        check_if_pkg_is_installed omi
+        if [ $? -eq 0 ]; then
+            pkg_upd $OMI_PKG omi
+            # It is acceptable that this fails due to the new omi being
+            # the same version (or less) than the one currently installed.
+            OMI_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Yeah, I guess we still need this (set to zero), since this doesn't include the version checking code. I imagine that installp installer will not install the same version.

File: installer/bundle/bundle_skel_AIX.sh (hunk 18 above)
Comment (jeffaco) [inline installer/bundle/bundle_skel_AIX.sh]: Wow, you even made this work on the various platforms. Great job!

I think I'm just more fundamentally lazy than you, which probably is not a good thing. :disappointed:

File: installer/bundle/bundle_skel_Linux.sh [hunk 22]
```diff
@@ -471,7 +394,7 @@ do
         --extract-binary)
             # hidden option, not part of usage
             # echo "  --extract-binary FILE  extract the binary to FILE."
-            tail $TAIL_CQUAL +${SCRIPT_LEN_PLUS_ONE} "${SCRIPT}" > "$2"
+            tail -n +${SCRIPT_LEN_PLUS_ONE} "${SCRIPT}" > "$2"
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: Nice touch (getting rid of the `$TAIL_CQUAL` thing).

File: installer/bundle/bundle_skel_Linux.sh [hunk 23]
```diff
@@ -690,8 +569,8 @@ fi
 #
 
 EXIT_STATUS=0
-SCX_EXIT_STATUS=0
 OMI_EXIT_STATUS=0
+SCX_OMI_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: Not quite sure why you renamed SCX_EXIT_STATUS to SCX_OMI_EXIT_STATUS. I suggest you keep it as `SCX_EXIT_STATUS` (consistent with other files, and clearer as well).

Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: Oh, I see why you renamed this. This is now combined due to bundling of operations. No, this makes perfect sense, keep it just the way it is!

Made me wonder what I did in OMS. I just called it KIT_STATUS (since I was installing a bunch of different things). I was fearful I got too lazy and didn't rename it to something meaningful! :disappointed:  But no, we're both good! :smile:

File: installer/bundle/bundle_skel_Linux.sh [hunk 24]
```diff
@@ -702,29 +581,21 @@ case "$installMode" in
...
         fi
 
-        pkg_add $OM_PKG scx
-        SCX_EXIT_STATUS=$?
+        pkg_add_list $OM_PKG scx
+
+        pkg_add ${ADD_PKG_QUEUE}
+        SCX_OMI_EXIT_STATUS=$?
 
         if [ $PROVIDER_ONLY -eq 0 ]; then
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: I don't think we actually create a provider-only kit anymore. You can keep it if you want, or you can delete it.

Originally, when planning OMS, the thought was to install SCX itself as part of a mega-bundle (i.e. check in the SCX installation shell bundle from the build into OMS itself, much like we have Apache/MySQL kits checked into SCXcore-osskits). But in the end, I didn't do it this way, so making these "provider-only" changes wasn't actually used.

Now, that said, I think there's support code for that in the configure script. So if you were to remove it, you'd need to remove it from configure, make, and here as well.

It's probably not worth it (remember, I'm fundamentally lazy), but I'll leave that up to you. I wonder if this sort of thing will come in useful later? Probably not, but who knows. I suspect that if it did come up later, it would be in some other form (like not include the MySQL/Apache bundles as part of the kit). But if we did that, it would just involve changing the configure and makefile scripts anyway.

File: installer/bundle/bundle_skel_Linux.sh [hunk 25]
```diff
@@ -749,19 +620,19 @@ case "$installMode" in
...
-                OMI_EXIT_STATUS=0
+                pkg_upd_list $OMI_PKG omi $?
             else
-                pkg_add $OMI_PKG omi
-                OMI_EXIT_STATUS=$?  
+                pkg_add_list $OMI_PKG omi
+                pkg_add ${ADD_PKG_QUEUE}
+                OMI_EXIT_STATUS=$?
             fi
         fi
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_Linux.sh]: I'm don't think you need to do this, but I'm not 100% sure. I'm not sure if the version checking code is "healthy" if a kit isn't installed at all, but I think that I made it healthy in the end.

I think that you can remove lines 620 and 621 (checking if package is installed), and then remove 624-628 (else case and calling pkg_add_list/pkg_add). If I remember properly, "shouldInstall_omi" says "yes" if OMI is not installed *OR* if OMI is installed but an older version. If this is the case, I know for certain that RPM could care less if you select -U (upgrade) to install a kit that isn't previously installed.

Doing this is probably worthwhile, as it gets rid of the notion of `OMI_EXIT_STATUS` vs. `SCX_OMI_EXIT_STATUS` (I think), and then you can just have `SCX_OMI_EXIT_STATUS`.

File: installer/bundle/bundle_skel_SunOS.sh [hunk 26]
```diff
@@ -0,0 +1,432 @@
...
+
+        # If the new version is greater than the previous, upgrade it. We expect at least 3 tokens in the version.
+        UPGRADE_PACKAGE=0
+        if [ $FV_1 -gt $IV_1 -o $FV_2 -gt $IV_2 -o  $FV_3 -gt $IV_3 ]; then
+            UPGRADE_PACKAGE=1
+        elif [ -n "$FV_4" -a -n "$IV_4" ]; then
+            if [ $FV_4 -gt $IV_4 ]; then
+                UPGRADE_PACKAGE=1
+            fi
+        fi
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_SunOS.sh]: Holy smokes, I had no idea this was here!

If it works, it works. But I wonder if it's worth while to bring the Linux version comparison code over here as well to deal with Sun's crappy installer?

On the bright side, with all these changes, you'd only focus on testing the Solaris bundle, so you'd get to immediately benefit from the changes! :smile:

It took me a while to get the version comparison code "right" for Linux, and frankly, I'm not 100% sure this is right here. But I'll leave that up to you in the end.

Comment (jeffaco) [inline installer/bundle/bundle_skel_SunOS.sh]: If you do bring over the Linux version comparison code, I'd like to see a review of just those files. Probably worth doing that separately (i.e. merge this set of changes, close out the Pull Request, then do merge of Linux version comparison code under a separate Pull Request).

File: installer/bundle/bundle_skel_SunOS.sh [hunk 27]
```diff
@@ -0,0 +1,432 @@
...
+            echo "ERROR: SCX package is already installed"
+            cleanup_and_exit 2
+        fi
+
+        check_if_pkg_is_installed omi
+        if [ $? -eq 0 ]; then
+            pkg_upd $OMI_PKG omi
+            # It is acceptable that this fails due to the new omi being
+            # the same version (or less) than the one currently installed.
+            OMI_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_SunOS.sh]: This this valid here? Notice how pkg_upd did some version checking. Given that, does this even make sense here?

I think we missed that earlier because it was hidden in a bunch of other if'ed code. But I'd think, given the version comparison code already in pkg_upd, that you would want to pick up the OMI_EXIT_STATUS and not ignore what happened.

File: installer/bundle/bundle_skel_SunOS.sh [hunk 28]
```diff
@@ -0,0 +1,432 @@
...
+        ;;
+
+    U)
+        echo "Updating cross-platform agent ..."
+        check_if_pkg_is_installed omi
+        if [ $? -eq 0 ]; then
+            pkg_upd $OMI_PKG omi
+            # It is acceptable that this fails due to the new omi being
+            # the same version (or less) than the one currently installed.
+            OMI_EXIT_STATUS=0
```
Comment (jeffaco) [inline installer/bundle/bundle_skel_SunOS.sh]: Ditto with above, I wouldn't think this makes sense here given that pkg_upd does version comparison. Depending on how robust that is, you may not even need to check if OMI is installed (although I'd guess it's not that robust). But definitely, if that code all works "more or less properly", I think we should be picking up the exit status code properly and using it to determine our exit status.

File: installer/bundle/create_bundle.sh [hunk 29]
```diff
@@ -155,8 +159,13 @@ OUTPUT_DIR=`(cd $2; pwd -P)`
...
 cd $INTERMEDIATE_DIR
 
+if [ ! -f $SOURCE_DIR/bundle_skel_${PLATFORM_TYPE}.sh ]
+then
+    echo "shell bundle skeleton file bundle_skel_${PLATFORM_TYPE}.sh does not exist in ${SOURCE_DIR}"
+    exit 1
+fi
 # Fetch the bundle skeleton file
-cp $SOURCE_DIR/bundle_skel.sh .
+cp $SOURCE_DIR/bundle_skel_${PLATFORM_TYPE}.sh ./bundle_skel.sh
```
Comment (jeffaco) [inline installer/bundle/create_bundle.sh]: Might be with pre-flighting this with a meaningful exit code.  Something like:

```
if [ ! -f $SOURCE_DIR/bundle_skel_${PLATFORM_TYPE}.sh]
then
    echo "Can't find source bundle file $SOURCE_DIR/bundle_skel_${PLATFORM_TYPE}.sh ..."
    exit 1
fi
```

Then do the cp operation at 168.

The concern is that if something was changed, it might be hard to figure out exactly why the create_bundle script is unhappy ...

**Comment by jeffaco** [comment]
Very nicely done. I wish these changes were in place years ago, it would have made our maintenance work a whole lot easier. You really did do a great job with this, it's so much cleaner and nicer, and it will pay off in spades with easier maintenance down the road.

If you make any further changes, verify your installation tests. If you took shortcuts in prior installation tests, do full testing now (after any final changes). You want to try hard not to commit something broken to master considering we're getting close to shipping the 2016 version.

**You have sign-off.** I did note that you have two separate commit messages. Please merge this into one meaningful commit message before merging to master. If you need instructions on how to do that, let me know, although I think it's documented (more or less) in ostc-docs.

Basically, assuming that your two commit messages are the two most recent changes in the `git log` output, then do:

```
git reset --soft HEAD^
git commit --amend
```

The `git commit --amend` should bring up an editor where you can correct your commit message.

The `git reset --soft HEAD^` command basically says "undo the last commit, but keep any source changes made in that commit staged". Thus, when you `git commit --amend`, the reset changes are merged with the prior changes.

File: installer/bundle/create_bundle.sh (hunk 29 above)
Comment (kapilash) [inline installer/bundle/create_bundle.sh]: The lines 162-166 are doing the same, no? 
I believe you want me to fix the error message (or unlikely, you missed the if block that I introduced earlier - you had mentioned this in an email earlier and I promptly introduced that)

**Comment by kapilash** [comment]
Thanks for all the guidance.  I've included the additional comments related to ```OMI_EXIT_STATUS``` and the changes needed in the sunos skeleton.

I squashed the commits and did thorough checking :
*  pbuild 
* upgrade (with and without ```--force```)
* install  

on all the pbuild platforms (as well as on an RHEL vm).

**Linux Comparison Code to SunOs skeleton** : I will send these changes in a separate PR. 


**Removing ```PROVIDER_ONLY``` for Linux** :  I will club this too with the above new PR. I will analyze the impact on ```make``` and ```configure``` before I do so.

File: installer/bundle/bundle_skel_Linux.sh (hunk 24 above)
Comment (kapilash) [inline installer/bundle/bundle_skel_Linux.sh]: Thanks for patiently explaining this. I understand it better now. I'll look into the impact on configure and make.  I will try to club the changes for these with the PR for SunOS skeleton.

File: installer/bundle/create_bundle.sh (hunk 29 above)
Comment (jeffaco) [inline installer/bundle/create_bundle.sh]: Whoops, somehow I missed that. My mistake, you're totally correct. Sorry 'bout that.
